### PR TITLE
Feature/unresolved resonances

### PR DESCRIPTION
**Pull request by jlconlin**

### CODE REVIEW

File: src/ENDFtk/resonanceParameters/Base.hpp [hunk 1]
```diff
@@ -7,76 +7,5 @@ class Base : private CONT {
...
-      }
-      break;
-      
-    default:
-      Log::error( "Invalid NRO value: {}. ", this->NRO() );
-      Log::info( "Valid NRO values are 0 or 1" );
-      throw 4;
-    }
-  }
+#include "ENDFtk/resonanceParameters/Base/src/ctor.hpp"
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/Base.hpp]: 👍

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp [hunk 2]
```diff
@@ -24,42 +24,6 @@ class SpecialCase {
     el( base.EL() ), eh( base.EH() ), spi( cont.C1() ), ap( cont.C2() ){
     assert( base.LRU() == 0 );
     assert( base.LRF() == 0 );
-    if( base.NRO() ){
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: why remove validation?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
+#define CATCH_CONFIG_MAIN
+
+#include "catch.hpp"
+#include "ENDFtk.hpp"
+
+using namespace njoy::ENDFtk;
+
+std::string LRF2() {
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp]: forward declare and put these definitions at the end of the file

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
+#define CATCH_CONFIG_MAIN
+
+#include "catch.hpp"
+#include "ENDFtk.hpp"
+
+using namespace njoy::ENDFtk;
+
+std::string LRF2(){
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp]: forward declare and move implementation to the end of the file

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/test/EnergyDependent.test.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
+#define CATCH_CONFIG_MAIN
+
+#include "catch.hpp"
+#include "ENDFtk.hpp"
+
+using namespace njoy::ENDFtk;
+
+std::string LRF2(){
+  return
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/test/EnergyDependent.test.cpp]: forward declare these and move them to the end of the file

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp [hunk 6]
```diff
@@ -0,0 +1,17 @@
+class LState {
+protected:
+  std::vector< LIST > lists;
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: smart field ordering

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+#include "ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/ctor.hpp"
+#include "ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp"
+
+public:
+  double AWRI() const { return this->awri; }
+  int L() const { return this->l; }
+
+  auto LISTs() const {
+    return ranges::make_iterator_range( this->lists.begin(), this->lists.end() );
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: nice!

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
+template< typename Iterator >
+static std::vector< LIST >
+readLists( int NJS,
+           Iterator& it, const Iterator& end, long& lineNumber,
+           int MAT, int MF, int MT ){
+  if ( NJS < 0 ){
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp]: empty lstates are possible for ENDF files with multiple isotopes for resolved. I'll double check for unresolved.

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    Log::error( "NJS required to >= 0" );
+    Log::error( "NJS value: {}", NJS );
+    throw std::exception();
+  }
+
+  std::vector< LIST > lists;
+  lists.reserve( NJS );
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp]: nice

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/test/LState.test.cpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
+#define CATCH_CONFIG_MAIN
+
+#include "catch.hpp"
+#include "ENDFtk.hpp"
+
+using namespace njoy::ENDFtk;
+
+std::string ENDF(){
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/test/LState.test.cpp]: forward declare and move implementation to end of file

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  int NAPS() const { return this->naps; }
+  
+  const auto& APE() const { return *(this->ape); }
+
+  double SPI() const { return this->spi; }
+  double AP() const { return this->ap; }
+  int LSSF() const { return this->lssf; }
+  int NLS() const { return this->lStates.size(); }
+
+  auto LStates() const {
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp]: not sure if counts as camel case or not, but reads fine

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/src/readLISTs.hpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
+template< typename Iterator >
+static std::vector< LIST >
+readLISTs( int NLS,
+           Iterator& it, const Iterator& end, long& lineNumber,
+           int MAT, int MF, int MT ){
+
+  if ( NLS < 1 ){
+    Log::error( "NLS required to be positive" );
+    Log::error( "NLS value: {}", NLS );
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/src/readLISTs.hpp]: error -> info

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readLStates.hpp [hunk 13]
```diff
@@ -0,0 +1,20 @@
+template< typename Iterator >
+static std::vector< LState >
+readLStates( int NLS,
+           Iterator& it, const Iterator& end, long& lineNumber,
+           int MAT, int MF, int MT ){
+
+  if ( NLS < 1 ){
+    Log::error( "NLS required to be positive" );
+    Log::error( "NLS value: {}", NLS );
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readLStates.hpp]: error -> info

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp [hunk 14]
```diff
@@ -0,0 +1,45 @@
+private:
+template< typename Iterator >
+EnergyDependentFissionWidths( const Base& base,
+                              std::optional<TAB1>&& APE,
+                              // const CONT& cont,
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp]: safe to remove?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+                              const LIST&& list,
+                              Iterator& it, const Iterator& end, long& lineNumber,
+                              int MAT, int MF, int MT ) 
+  try:
+    Unresolved( base, std::move( APE ),
+                CONT( list.C1(), list.C2(), 
+                      list.L1(), list.L2(), 
+                      list.NPL(), list.N2() ),
+                // readLStates( CONT( list.C1(), list.C2(),
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp]: safe to remove?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+template< typename Iterator >
+EnergyDependentFissionWidths( const Base& base,
+                              std::optional<TAB1>&& APE,
+                              Iterator& it, const Iterator& end, long& lineNumber,
+                              int MAT, int MF, int MT ) :
+  EnergyDependentFissionWidths( base, 
+                                std::move( APE ),
+                                // CONT( it, end, lineNumber, MAT, MF, MT ),
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp]: safe to remove?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp [hunk 17]
```diff
@@ -0,0 +1,14 @@
+class EnergyDependentFissionWidths : public Unresolved {
+protected:
+#include "ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp"
+
+  LIST list;
+  int ne; // Number of energy points at which there are energy-dependent widths
+
+public:
+
+  double NE() const { return this->ne; }
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp]: isn't this list.NPL()?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+  auto energyLIST() const { return this->list; }
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp]: return by const reference

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/src/readLISTs.hpp [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    throw std::exception();
+  }
+  std::vector< LIST > lists;
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/src/readLISTs.hpp]: if you move this to the first line you can get rvo

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+readLists( int NJS,
+           Iterator& it, const Iterator& end, long& lineNumber,
+           int MAT, int MF, int MT ){
+  if ( NJS < 0 ){
+    Log::error( "NJS required to >= 0" );
+    Log::error( "NJS value: {}", NJS );
+    throw std::exception();
+  }
+
+  std::vector< LIST > lists;
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp]: if you move this to the first line you can get rvo

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp [hunk 21]
```diff
@@ -0,0 +1,9 @@
+template< typename Iterator >
+static auto
+readAPE( const Base& base,
+         Iterator& it, const Iterator& end, long& lineNumber,
+         int MAT, int MF, int MT ){
+  return ( base.NRO() ) ?
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp]: i like

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readLStates.hpp [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+    throw std::exception();
+  }
+  std::vector< LState > lists;
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readLStates.hpp]: if you move this to the first statement, you can get rvo

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp [hunk 23]
```diff
@@ -0,0 +1,14 @@
+class EnergyDependentFissionWidths : public Unresolved {
+protected:
+#include "ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/ctor.hpp"
+
+  LIST list;
+  int ne; // Number of energy points at which there are energy-dependent widths
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp]: see comment on NE method

File: src/ENDFtk/resonanceParameters/Isotope/src/readRanges.hpp [hunk 24]
```diff
@@ -38,9 +56,10 @@ readRange( Iterator& it, const Iterator& end, long& lineNumber,
 
 template< typename Iterator >
 static std::vector< EnergyRange >
-readRanges( int NER,
+readRanges( const CONT& cont, 
             Iterator& it, const Iterator& end, long& lineNumber,
             int MAT, int MF, int MT ){
+  auto NER = cont.N1();
   std::vector< EnergyRange > ranges;
```
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/Isotope/src/readRanges.hpp]: if you swap this and line 62, you get rvo

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp (hunk 2 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: were they redundant?

Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: They were redundant with base.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp (hunk 3 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp]: What's wrong with having them here?

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp (hunk 6 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: huh?

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp (hunk 8 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState/src/readLists.hpp]: This allows for empty l-states (i.e., NJS=0). There was nothing I could find that explicitly prevented or forbad NJS=0 so I assumed it was the same as or the resolved resonance region.

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp (hunk 11 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp]: Uppercase `L` is used in the manual, which is explicitly stated as the preferred naming scheme in the Style Guide. (See [https://njoy.github.io/DevelopersGuide/Style/Code.html]() section 4 exception.)

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp (hunk 17 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp]: yep

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp (hunk 21 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp]: Of course you do; you wrote it!

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp (hunk 11 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp]: that's fine

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp (hunk 6 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: Happy coincidence then. This ordering of fields minimizes the memory footprint of the class.

Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: Can you be more specific so I can do this again in the future? Why is the order that I've set optimal?

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp (hunk 2 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: I don't think it is. Base allows for non-zero NRO values, whereas the special case doesn't, right?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp (hunk 17 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths.hpp]: so we can remove the `ne` field then

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp (hunk 2 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: That's true. The `SpecialCase` requires that `NRO`=0.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp (hunk 3 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp]: a user has to search through 2/3 of the file before getting to the tests they were interested in looking at

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp (hunk 11 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp]: my note was regarding `Lstates` vs `LStates`

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp (hunk 21 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/src/readAPE.hpp]: Everybody likes their own brand I guess

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp (hunk 11 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved.hpp]: to be clear, I'm not asking for any changes here

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp (hunk 2 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: In addition to the `L1`, `L2`, `N1`, `N2` tests

File: src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp (hunk 6 above)
Comment (apmccartney) [inline src/ENDFtk/resonanceParameters/unresolved/Unresolved/LState.hpp]: We might need a white board, but happy to. It has to do with the alignment requirements of the types embedded types, and how that can require gaps of memory between member variables. As a general rule of thumb, ordering fields from biggest to smallest (usually) gets you pretty close to optimal

File: src/ENDFtk/resonanceParameters/SpecialCase.hpp (hunk 2 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/SpecialCase.hpp]: Yep. This is now fixed.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp (hunk 3 above)
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/test/EnergyIndependent.test.cpp]: code folding. Perhaps emacs doesn't do that? ;)

**Comment by apmccartney** [comment]
Excellent work Jeremy

**Pull request merged**